### PR TITLE
PBOT-102: delete link from project action

### PR DESCRIPTION
**Pull request by colbyallenc**
/project delete deletes item from db
deleteLinkFromProjectAction deletes a link from the db and sends response to slack
PBOT-102

### CODE REVIEW

File: lib/actions/deleteLinkFromProjectAction.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+const ProjectLink = require('../projectLink');
+
+class DeleteLinkFromProjectAction {
+  constructor(projectLink = ProjectLink) {
+    this.projectLink = projectLink;
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  parseParams(userInputString = '') {
+    assert(typeof userInputString === 'string');
```
Comment (nrambeck) [inline lib/actions/deleteLinkFromProjectAction.js]: Asserts are usually used just for testing. I believe we carried this over from a previous action. Do you know what it's purpose is here?

File: lib/projectLink.js [hunk 2]
```diff
@@ -9,6 +9,10 @@ class ProjectLink {
     await db.addLink(this.project, this.label, this.url);
   }
 
+  async delete(db) {
+    await db.deleteLink(this.project, this.label, this.url);
```
Comment (nrambeck) [inline lib/projectLink.js]: The final parameter (url) is unnecessary here as it is not actually used by the `deleteLink` method.

File: lib/actions/deleteLinkFromProjectResponse.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class DeleteLinkFromProjectResponse {
+  constructor(label, link) {
+    this.label = label;
+    this.link = link;
+  }
+
+  /* eslint-disable-next-line class-methods-use-this */
+  sendToBot(bot, message) {
+    return bot.reply(message, 'Link Deleted.');
```
Comment (nrambeck) [inline lib/actions/deleteLinkFromProjectResponse.js]: I recommend being more explicit about which link was deleted in this response.

File: lib/actions/deleteLinkFromProjectAction.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    const parsedUserInput = userInputString.split(' ');
+
+    if (parsedUserInput.length < 2) {
+      throw new TypeError('invalid input');
+    }
+    const linkParams = { label: parsedUserInput[1], link: parsedUserInput[2] };
```
Comment (nrambeck) [inline lib/actions/deleteLinkFromProjectAction.js]: We will only need the label of the link and not the URL, in order to delete. So we would not expect nor need a link parameter, right?

File: lib/actions/deleteLinkFromProjectAction.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    return linkParams;
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  async execute(params, context) {
+    const paramsObj = this.parseParams(params);
+    // eslint-disable-next-line new-cap
+    const projectLink = new this.projectLink(context.project, paramsObj.label, paramsObj.link);
```
Comment (nrambeck) [inline lib/actions/deleteLinkFromProjectAction.js]: This is probably working, but it isn't really doing what we want it to do. Currently, this line is creating a brand new instance of a new link object from the parameters provided by the user. However, we want to find an existing projectLink record and create an instance from that record. That means we should have some kind of find method to find a project link based on the project ID and the label. Then once we have that instance, we can use the delete method on that instance. Also, using a find method will allow us to change our response to the user, if the label they provided doesn't match any links in the database.

File: lib/actions/deleteLinkFromProjectAction.js (hunk 1 above)
Comment (colbyallenc) [reply lib/actions/deleteLinkFromProjectAction.js]: This is a way that we can test to assert that the type of input from the user is a string. 
It's easy to test and just another way. We don't have to use this.

File: lib/projectLink.js [hunk 6]
```diff
@@ -10,13 +10,24 @@ class ProjectLink {
   }
 
   async delete(db) {
-    await db.deleteLink(this.project, this.label, this.url);
+    await db.deleteLink(this.project, this.label);
+    console.log('deleted');
```
Comment (colbyallenc) [inline lib/projectLink.js]: I have to delete this console.log... i didnt catch it!

File: lib/dbDriver/mysql.js [hunk 7]
```diff
@@ -25,6 +25,15 @@ class MySQLDriver {
       .from('project_links')
       .where({ project_id: project.project_id });
   }
+
+  // eslint-disable-next-line class-methods-use-this
+  async getProjectLabelById(project, label) {
```
Comment (nrambeck) [inline lib/dbDriver/mysql.js]: I would call this `getProjectLinkByLabel`.

File: lib/projectLink.js [hunk 8]
```diff
@@ -10,13 +10,24 @@ class ProjectLink {
...
   }
 
   static async findByProject(project, db) {
     const links = await db.getAllLinksByProject(project);
     return links.map(link => new ProjectLink({ project: link.project_id }, link.label, link.url));
   }
+
+  static async findProjectLabelById(project, db, label) {
```
Comment (nrambeck) [inline lib/projectLink.js]: I would call this `findByProjectAndLabel`

File: lib/actions/deleteLinkFromProjectResponse.js [hunk 9]
```diff
@@ -6,7 +6,7 @@ class DeleteLinkFromProjectResponse {
 
   /* eslint-disable-next-line class-methods-use-this */
   sendToBot(bot, message) {
-    return bot.reply(message, 'Link Deleted.');
+    return bot.reply(message, `${this.label} deleted :white_check_mark:`);
```
Comment (nrambeck) [inline lib/actions/deleteLinkFromProjectResponse.js]: Nice and 👍 for the emoji!

File: lib/projectLink.js (hunk 6 above)
Comment (nrambeck) [reply lib/projectLink.js]: Don't forge to remove this.

**Pull request merged**